test_client.py

```diff
@@ -1,15 +1,45 @@
 import echo_client
 
 
-def test_basic_entry():
-    assert echo_client.start_client("Hello") == "Received message: Hello"
+def test_HTTP_bad_response_error():
+    msg = echo_client.start_client("""
+            Hello!
+            """)
+    error_response = "HTTP/1.1 400 ERROR\r\n"
+    headers = "Content-Type: text/plain\r\n"
+    body = "ERROR 400, BAD REQUEST\r\n"
+    response = "{}{}{}".format(error_response, headers, body)
+    assert msg == response
 
 
-def test_long_entry():
-    msg = echo_client.start_client("This message is longer than buffer")
-    assert msg == "Received message: This message is longer than buffer"
+def test_HTTP_okay():
+    msg = echo_client.start_client("""
+            GET /fake/uri HTTP/1.1\r\nHost: localhost:50001\r\n
+            """)
+    okay_response = "HTTP/1.1 200 OK\r\n"
+    headers = "Content-Type: text/plain\r\n"
+    body = "you requested: /fake/uri"
+    response = "{}{}{}\r\n".format(okay_response, headers, body)
+    assert msg == response
 
 
-def test_unicode_entry():
-    msg = echo_client.start_client(unicode('Hello'))
-    assert msg == "Received message: Hello"
+def test_HTTP_method_error():
+    msg = echo_client.start_client("""
+            POST /path/to/index.html HTTP/1.1\r\nHost: localhost:50001\r\n
+            """)
+    error_response = "HTTP/1.1 405 ERROR\r\n"
+    headers = "Content-Type: text/plain\r\n"
+    body = "ERROR 405, POST METHOD NOT ALLOWED\r\n"
+    response = "{}{}{}".format(error_response, headers, body)
+    assert msg == response
+
+
+def test_HTTP_version_error():
+    msg = echo_client.start_client("""
+            GET /path/to/index.html HTTP/1.0\r\nHost: localhost:50001\r\n
+            """)
+    error_response = "HTTP/1.1 505 ERROR\r\n"
+    headers = "Content-Type: text/plain\r\n"
+    body = "ERROR 505, HTTP/1.0 NOT SUPPORTED\r\n"
+    response = "{}{}{}".format(error_response, headers, body)
+    assert msg == response
```

Comment:
You have written functional tests that address the server as a black box.  That's a valid testing strategy, but it doesn't get at whether individual functions are working properly.  You should also have unit tests that verify that when you call (for example) the `parse_request` method, you get a URI or you get errors if the method or protocol are wrong.